l10n_es_aeat_mod390_vat_prorrate/models/mod390.py

```diff
@@ -0,0 +1,139 @@
+                    raise exceptions.ValidationError(
+                        _('Sum of prorrate %  cannot be 0')
+                    )
+                elif total_percent > 100:
+                    raise exceptions.ValidationError(
+                        _('Sum of prorrate % cannot exceed 100')
+                    )
+
+    @api.multi
+    def check_prorrate_ids_amount_valid(self):
+        for item in self:
+            if item.vat_prorrate_type != 'none':
+                if item.prorrate_ids:
+                    amount = sum(x['amount'] for x in item.prorrate_ids)
+                    if amount == 0:
+                        raise exceptions.ValidationError(
+                            _('Total amount of the operations cannot be 0')
+                        )
+
+    @api.multi
+    def check_prorrate_ids_total_deduction_valid(self):
+        for item in self:
+            if item.prorrate_ids:
+                amount = sum(x['amount'] for x in item.prorrate_ids)
+                key = 'amount_allow_deduction'
+                amount_deduction = sum(x[key] for x in item.prorrate_ids)
+                if amount_deduction == 0:
+                    raise exceptions.ValidationError(
+                        _(
+                            'Total amount of the operations with the '
+                            'right to deduction cannot be 0'
+                        )
+                    )
+                elif amount_deduction > amount:
+                    raise exceptions.ValidationError(
+                        _(
+                            'Total amount of the operations with the '
+                            'right to deduct cannot exceed the amount of '
+                            'the operations'
+                        )
+                    )
+
+    @api.multi
+    def button_confirm(self):
+        for item in self:
+            if item.prorrate_ids:
+                if item.vat_prorrate_type != 'none':
+                    item.vat_prorrate_type = 'none'
+
+                item.check_prorrate_ids_percent_valid()
+                item.check_prorrate_ids_total_deduction_valid()
+
+            item.check_prorrate_ids_amount_valid()
+
+        return super(L10nEsAeatMod390Report, self).button_confirm()
+
+    @api.multi
+    @api.onchange('vat_prorrate_type')
+    def _onchange_vat_prorrate_types(self):
+        for item in self:
+            if item.vat_prorrate_type == 'none':
+                item.vat_prorrate_percent = 0
+
+    @api.constrains('vat_prorrate_percent')
+    def check_vat_prorrate_percent(self):
+        if self.vat_prorrate_type != 'none':
+            if not (0 < self.vat_prorrate_percent <= 100):
+                raise exceptions.ValidationError(
+                    _('VAT prorrate percent must be between 0.01 and 100')
+                )
+
+    @api.multi
+    def _prepare_tax_line_vals(self, map_line):
+        res = super(L10nEsAeatMod390Report, self)._prepare_tax_line_vals(
```

Comment:
```suggestion
        res = super()._prepare_tax_line_vals(
```
Is self.ensure_one () necessary? If it is not, I think it would be good to add a comment to avoid confusion.

Reply:
IMP thi change is possible, but simillar addon https://github.com/OCA/l10n-spain/blob/12.0/l10n_es_aeat_vat_prorrate/models/mod303.py#L101 contain exactly same code.

Reply:
It is not mandatory, but it is usually written like this, just `super()....`

Reply:
En Python 2 era necesario que fuera así la llamada del super. Ahora en P3 no, así que se puede acortar. Sobre lo del `self.ensure_one()`, eso solo se pone en el método padre normalmente, a no ser que quieras denotar también en tu método que se está manejando un registro solo.